### PR TITLE
Olympus 3

### PR DESCRIPTION
**Pull request by JoramD**
**When merged this pull request will:**

Added:
- Medical functionality for Medic course.
- Vehicle spawning function for Demolitions course.
- Heal command for all players.
- Personal arsenals at every teleport location.
- Feature overview.

Changed:
- Moved to Stratis from Hebontes. (Mostly performance reasons, Looking inland on Hebontes results in pretty severe frame drops. It's also far bigger than we need for the usage purposes.)
- Less focused on placed objects and more on natural terrain.
- Demolitions course is now less focused on EOD aspects.
- Basic Rifle, Pistol, Grenades & Launchers now form up "Weapons Handling" in one area.
- Rifle & Pistol now have an additional CQB course requirement.
- Spawn Area now has targets & dueling target range.
- Launcher target reset now waits 2 seconds before respawning targets.
- Launcher targets are now more in parity with the kind of launcher being used.
- MOUT course now has vehicle entry option.
- Marksman course no longer focused on range and more on target acquisition and engagements.
- Whiteboards now feature the related document and minimum rank. (Weapons Handling is the exception)
- Rifle & Pistol ranges now have an additional 10% extra time to complete them.
- Personal arsenals are grey boxes, Full arsenals are black boxes.
- Personal arsenal loads data on opening arsenal, not on player joining.
- Improved environmentControl and added overcast/ran actions.

Removed:
- Advanced Rifle & Pistol courses.
- Advanced Marksman functionality.

### CODE REVIEW

**Review by Mike-MF — APPROVED** [review]
Approved by all BoD members.

Waiting on merge until release time for any potential upcoming changes.

**Comment by jonpas** [comment]
> Moved to Stratis from Hebontes.

Might want to write down why. :)

**Comment by Mike-MF** [comment]
Done

File: Olympus.Stratis/description.ext [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// Description.ext for Olympus 2
+// Description.ext for Olympus 3.0
```
Comment (jonpas) [inline Olympus.Stratis/description.ext]: Are you going to change this number every time a minor update releases?

File: Olympus.Stratis/functions/fn_healCommand.sqf [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+ * None
+ */
+#include "..\script_component.hpp"
+
+["heal",
+    {
+        [ace_player] call ACEFUNC(medical_treatment,fullHealLocal);
+        ["ace_common_systemChatGlobal", "[TAC] Healed", ace_player] call CBA_fnc_targetEvent;
+    }, "all"
+] call CBA_fnc_registerChatCommand;
```
Comment (jonpas) [inline Olympus.Stratis/functions/fn_healCommand.sqf]:
```
["heal", {
    [ace_player] call ACEFUNC(medical_treatment,fullHealLocal);
    ["ace_common_systemChatGlobal", "[TAC] Healed", ace_player] call CBA_fnc_targetEvent;
}, "all", ] call CBA_fnc_registerChatCommand;
```

File: Olympus.Stratis/functions/fn_healCommand.sqf [hunk 3]
```diff
@@ -0,0 +1,15 @@
+/*
+ * Author: Tyrone
+ * Enables healing via #heal command in chat.
+ *
+ * Return Value:
+ * None
+ */
+#include "..\script_component.hpp"
+
+["heal",
```
Comment (jonpas) [inline Olympus.Stratis/functions/fn_healCommand.sqf]:
```suggestion
["tac-heal",
```
Avoid conflicts.

File: Olympus.Stratis/functions/fn_healCommand.sqf [hunk 4]
```diff
@@ -0,0 +1,15 @@
+/*
+ * Author: Tyrone
+ * Enables healing via #heal command in chat.
```
Comment (jonpas) [inline Olympus.Stratis/functions/fn_healCommand.sqf]:
```suggestion
 * Enables healing via #tac-heal command in chat.
```

File: Olympus.Stratis/functions/fn_healCommand.sqf [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+ *
```
Comment (jonpas) [inline Olympus.Stratis/functions/fn_healCommand.sqf]: Where are arguments and example?

File: Olympus.Stratis/functions/fn_personalArsenal.sqf [hunk 6]
```diff
@@ -60,16 +60,16 @@ if (_success) then {
         "",
         {
             params ["_target", "_player"];
-            [_target, _player, false] call ACEFUNC(arsenal,openBox);
+            [_target, _player, false] call ace_arsenal_fnc_openBox;
```
Comment (jonpas) [inline Olympus.Stratis/functions/fn_personalArsenal.sqf]: What was wrong?

File: Olympus.Stratis/functions/fn_personalArsenal.sqf [hunk 7]
```diff
@@ -60,16 +60,16 @@ if (_success) then {
...
 
     // Add custom arsenal action to object
-    [_crate, 0, ["ACE_MainActions"], _action] call call ACEFUNC(interact_menu,addActionToObject);
+    [_crate, 0, ["ACE_MainActions"], _action] call ACEFUNC(interact_menu,addActionToObject);
 
     INFO_1("Personal Arsenal loaded successfuly.");
 } else {
     ERROR_2("Failed to load available items for personalArsenal (Name: %1 - UID: %2)!",profileName,getPlayerUID _player);
-    ["Could not load available items for Personal Arsenal", 1.5, ace_player, 10] call ACEFUNC(common,displayTextStructured);
+    ["Could not load available items for Personal Arsenal", 1.5, ace_player, 10] call ace_common_fnc_displayTextStructured
```
Comment (jonpas) [inline Olympus.Stratis/functions/fn_personalArsenal.sqf]: Same.

File: Olympus.Stratis/functions/fn_rangeMachineGunTargetGroups.sqf [hunk 8]
```diff
@@ -0,0 +1,69 @@
+/*
+ * Author: Rory, Jonpas
+ * Allows for the instructor to choose what group of targets pop up for basic LMG range
```
Comment (jonpas) [inline Olympus.Stratis/functions/fn_rangeMachineGunTargetGroups.sqf]:
```suggestion
 * Allows for the instructor to choose what group of targets pop up for basic LMG range.
```

File: Olympus.Stratis/functions/fn_rangeMachineGunTargetGroups.sqf [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+ *
+ * Arguments:
+ * 0: Interaction object.
```
Comment (jonpas) [inline Olympus.Stratis/functions/fn_rangeMachineGunTargetGroups.sqf]:
```suggestion
 * 0: Controller <OBJECT>
```

File: Olympus.Stratis/functions/fn_rangeMachineGunTargetGroups.sqf [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+ *
+ * Return Value:
+ * None
+ *
+ * Example:
+ * [this] call TAC_Olympus_fnc_machineGunTargetGroups;
```
Comment (jonpas) [inline Olympus.Stratis/functions/fn_rangeMachineGunTargetGroups.sqf]:
```suggestion
 * [this] call TAC_Olympus_fnc_machineGunTargetGroups
```
Missing `range` in the name.

File: Olympus.Stratis/functions/fn_rangeMarksmanTargetGroups.sqf [hunk 11]
```diff
@@ -1,6 +1,6 @@
 /*
  * Author: Rory, Jonpas
- * Allows for instructor to control pop up targets on Driving course
+ * Allows for the instructor to choose what group of targets pop up for the Marksman range
```
Comment (jonpas) [inline Olympus.Stratis/functions/fn_rangeMarksmanTargetGroups.sqf]:
```suggestion
 * Allows for the instructor to choose what group of targets pop up for the Marksman range.
```

File: Olympus.Stratis/functions/fn_rangeMarksmanTargetGroups.sqf [hunk 12]
```diff
@@ -1,6 +1,6 @@
...
  *
  * Arguments:
  * 0: Interaction object.
```
Comment (jonpas) [inline Olympus.Stratis/functions/fn_rangeMarksmanTargetGroups.sqf]:
```suggestion
 * 0: Controller <OBJECT>
```

File: Olympus.Stratis/functions/fn_rangeMarksmanTargetGroups.sqf [hunk 13]
```diff
@@ -9,12 +9,12 @@
  * None
  *
  * Example:
- * [this] call TAC_Olympus_fnc_rangeDrivingCourseGroups;
+ * [this] call TAC_Olympus_fnc_MarksmanTargetGroups;
```
Comment (jonpas) [inline Olympus.Stratis/functions/fn_rangeMarksmanTargetGroups.sqf]:
```suggestion
 * [this] call TAC_Olympus_fnc_marksmanTargetGroups
```
It also still has the `range` in the name.

File: Olympus.Stratis/functions/fn_resetDamagedObjects.sqf [hunk 14]
```diff
@@ -29,15 +29,16 @@ private _action = [
                 private _type = typeOf _x;
                 deleteVehicle _x;
 
-                // Run this in the next frame
+                // Waits 2s then executes.
```
Comment (jonpas) [inline Olympus.Stratis/functions/fn_resetDamagedObjects.sqf]:
```suggestion
                // Run after 2 seconds
```
Obviously it executes, that's a redundant comment.

Why was this changed to 2 seconds though?

File: Olympus.Stratis/functions/fn_spawnDemolitionsVehicle.sqf [hunk 15]
```diff
@@ -0,0 +1,99 @@
+/*
+ * Author: JoramD
+ * Spawn a vehicle from pre-defined list.
+ *
+ * Arguments:
+ * 0: Controller <OBJECT>
+ * 1: Spawn Position <ANY>
```
Comment (jonpas) [inline Olympus.Stratis/functions/fn_spawnDemolitionsVehicle.sqf]: I am about 99% sure this isn't `ANY`.

File: Olympus.Stratis/functions/fn_spawnDemolitionsVehicle.sqf [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+ * Return Value:
+ * None
+ *
+ * Example:
+ * [this, spawnPos] call TAC_Olympus_fnc_engineerCourse
+ */
+#include "..\script_component.hpp"
+
+params ["_controller", "_spawnPos"];
+
```
Comment (jonpas) [inline Olympus.Stratis/functions/fn_spawnDemolitionsVehicle.sqf:18]:
```suggestion
```

File: Olympus.Stratis/functions/fn_spawnDemolitionsVehicle.sqf [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+private _spawnAction = [
+    QGVAR(SpawnVehicle),
+    "Spawn Vehicle",
+    "",
+    {},
+    {
+        (_this select 2) params ["_controller", "_spawnPos"];
+
+        private _spawnedVehicle = _controller getVariable [QGVAR(spawnedVehicle), []];
+
```
Comment (jonpas) [inline Olympus.Stratis/functions/fn_spawnDemolitionsVehicle.sqf]:
```suggestion
```

File: Olympus.Stratis/initPlayerLocal.sqf [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+ * Author: Theseus Incorporated
+ *
+ * initPlayerLocal.sqf
+ * Initialization of scripts local to player
+ * Executed locally when player joins mission (includes both mission start and JIP)
+ * Executed after initServer and before initPlayerServer and init.
+ *
+ * Arguments: [player:Object, didJIP:Boolean]
+ *
+ */
```
Comment (jonpas) [inline Olympus.Stratis/initPlayerLocal.sqf:17]: This header should be changed to ACE standard.
Should be standardized with template as well.

File: Olympus.Stratis/initPlayerLocal.sqf [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+
+#include "script_component.hpp"
```
Comment (jonpas) [inline Olympus.Stratis/initPlayerLocal.sqf]: All `script_component.hpp` should go to top of file (above header).

File: Olympus.Stratis/initPlayerLocal.sqf [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+
+[] call FUNC(healCommand);
```
Comment (jonpas) [inline Olympus.Stratis/initPlayerLocal.sqf]: Might as well name it `addChatCommands`, they are usually super short. Also no need for empty array parameter if it doesn't need any.

File: Olympus.Stratis/initPlayerLocal.sqf [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+[
+    [
+        ["OLYMPUS", "<t size = '1.5' underline = '1'>%1</t><br/>"],
+        ["Theseus Incorporated"],
+        ["Stratis", "<t size = '1' font='puristaMedium'>%1</t>", 70]
+    ], 1, 0.75, "<t align = 'center' shadow = '1' size = '1.0'>%1</t>"
+] spawn BIS_fnc_typeText;
+
+// Sets default loadout to joining players
+[_this select 0] call FUNC(setBasicLoadout);
```
Comment (jonpas) [inline Olympus.Stratis/initPlayerLocal.sqf]: Do a `params ["_player", "_didJIP"];` on top and use it here.
```suggestion
[_player] call FUNC(setBasicLoadout);
```
Refer to template.

**Review by jonpas — CHANGES_REQUESTED** [review]

File: Olympus.Stratis/functions/fn_resetDamagedObjects.sqf (hunk 14 above)
Comment (JoramD) [reply Olympus.Stratis/functions/fn_resetDamagedObjects.sqf]: There were still vehicles flying off into outer space, removed the comment.

File: Olympus.Stratis/functions/fn_addChatCommands.sqf [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+                if ([_name, name _x] call BIS_fnc_inString) then {
+                    _matches pushBack _x;
+                };
+            } forEach (call CBA_fnc_players);
+
+            if (count _matches == 1) exitWith {_matches select 0};
+            objNull
+        };
+
+        private _foundPlayer = _name call _fnc_findMatch;
```
Comment (jonpas) [inline Olympus.Stratis/functions/fn_addChatCommands.sqf]:
```suggestion
        private _foundPlayers = (call CBA_fnc_players) select {_name in (name _x)};
        // proceed to check if multiple, then tell user to specify in more detail, otherwise take the only (first) match
```

**Review by jonpas — COMMENTED** [review]
Better name search (1-liner!).

File: Olympus.Stratis/functions/fn_environmentControl.sqf [hunk 23]
```diff
@@ -15,143 +15,255 @@
...
+        private _actions = [];
         {
-            (_this select 2) params ["_time"];
-            [QGVAR(setTime), [_time]] call CBA_fnc_globalEvent;
-        },
-        {true},
-        {},
-        [_time]
-    ] call ACEFUNC(interact_menu,createAction);
+            _x params ["_time", "_timeText"];
```
Comment (jonpas) [inline Olympus.Stratis/functions/fn_environmentControl.sqf]: You can just get time text from time:
```sqf
private _timeText = format ["%100", str _time];
if (count _timeText < 4) {
    _timeText = ["0", _timeText] joinString "";
};
```
If wanted.

File: Olympus.Stratis/functions/fn_environmentControl.sqf [hunk 24]
```diff
@@ -25,23 +25,28 @@ private _timeAction = [
...
-            [7, "0700"],
-            [9, "0900"],
-            [11, "1100"],
-            [13, "1300"],
-            [15, "1500"],
-            [17, "1700"],
-            [19, "1900"],
-            [21, "2100"],
-            [23, "2300"]
+            [1],
```
Comment (jonpas) [inline Olympus.Stratis/functions/fn_environmentControl.sqf]: You don't need an array in an array anymore. :D

Actually you don't need to define that at all. Just do:
```sqf
for "_time" from 1 to 23 step 2 do {
    //_time
};
```

**Review by jonpas — APPROVED** [review]

File: Olympus.Stratis/initPlayerLocal.sqf [hunk 25]
```diff
@@ -31,3 +31,6 @@ call FUNC(addChatCommands);
 
 // Sets default loadout to joining players
 [_player] call FUNC(setBasicLoadout);
+
+// Feature overview
+[player] call FUNC(featureOverview);
```
Comment (jonpas) [inline Olympus.Stratis/initPlayerLocal.sqf:36]: You already have player in a variable.
